cmd: get ep with labels, pod / container name, ID

ianvernon opened the pull request:
cmd: get ep with labels, pod / container name, ID
    
* daemon: add handlers for new /endpoints API, update endpoint structures with container ID / pod name
* pkg/endpoint: support new API models
* pkg/endpointmanager: add pod name, container name entries as keys mapping to endpoints
* pkg/labels: change Orchestration labels -> OrchestrationIdentity labels (labels used in computing a security identity) and OrchrestationInfo labels (labels not used in computing a security identity)
    
`cilium endpoint get` now supports the following:
* `cilium endpoint get -l <set of labels>`
* `cilium endpoint get <eID, pod name, container name, etc.>`

api: add pod / container name, labels for GET ep
    
* add new fields to Endpoint (pod-name, container-name)
* GET /endpoint now takes an optional array of labels to return a list of endpoints whose labels match the provided array.
    
Signed-off by: Ian Vernon <ian@covalent.io>

tgraf commented on daemon/docker_watcher.go:
```diff
@@ -249,6 +255,21 @@ func (d *Daemon) handleCreateContainer(id string, retry bool) {
 			}
 		}
 
+		if ep != nil {
```
Just move this to below the `ep ==nil` condition so you don't have to check for `ep != nil`

ianvernon replied:
Done.

tgraf commented on daemon/docker_watcher.go:
```diff
@@ -249,6 +255,21 @@ func (d *Daemon) handleCreateContainer(id string, retry bool) {
+			if d.conf.IsK8sEnabled() {
+				if dockerContainer.Config != nil {
+					podNamespace := k8sDockerLbls.GetPodNamespace(dockerContainer.Config.Labels)
+					podName := k8sDockerLbls.GetPodName(dockerContainer.Config.Labels)
```
Does this trigger another API call to the API server? If so, can we restructure the code to retrieve this when we already fetch them via `getFilteredLabels()`?

ianvernon replied:
These calls to `GetPodNamespace` and `GetPodName` are lookups into the map `dockerContainer.Config.Labels` for `io.kubernetes.pod.name` and `io.kubernetes.pod.namespace` - so they are not calls to the API server.

tgraf commented on daemon/endpoint.go:
```diff
@@ -99,6 +103,48 @@ func (h *getEndpointID) Handle(params GetEndpointIDParams) middleware.Responder
+		args.Add("label", v)
+	}
+
+	containers, err := h.d.dockerClient.ContainerList(ctx.Background(), types.ContainerListOptions{
```
This is not a good idea. This will potentially block this API call for a long time. We should fetch and store the docker labels with the endpoint when we receive the create event and then just look for local endpoints which match the labels.

ianvernon replied:
Done.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -174,32 +174,34 @@ const (
-	State            string
+	ID                  uint16       // Endpoint ID.
+	Mutex               sync.RWMutex // Protects all variables from this structure below this line
+	DockerContainerName string       // Docker container name.
```
Can we call this `ContainerName`? We will rename the `DockerID` to `ContainerID` as well later.

ianvernon replied:
Done.

tgraf commented on pkg/endpoint/id.go:
```diff
@@ -47,6 +49,9 @@ func NewID(prefix PrefixType, id string) string {
 	if s := strings.Split(id, ":"); len(s) == 2 {
 		return PrefixType(s[0]), s[1]
+	} else if len(s) == 3 {
+		// PodNamePrefix case.
```
Can you add an example here on how the id looks like to make it clear?

ianvernon replied:
Done.

tgraf reviewed (changes requested):
This will be awesome!!! 👍

aanm commented on api/v1/openapi.yaml:
```diff
@@ -224,6 +224,27 @@ paths:
           x-go-name: Failed
           schema:
             "$ref": "#/definitions/Error"
+  "/endpoints/":
```
Why not extending the `/endpoint/` on line 170?

ianvernon replied:
`/endpoint/` returns one endpoint, while `/endpoints/` returns an array of endpoints. I'm open to moving this under `/endpoint/`, though and changing `/endpoint/` to return a list of endpoints as well.

aanm replied:
```
   "/endpoint":
     get:
       summary: Get list of all endpoints
       description: |
         Returns an array of all local endpoints.
```

ianvernon replied:
You're right :) I've made this change as part of my most recent commit.

aanm commented on cilium/cmd/endpoint_get.go:
```diff
@@ -18,28 +18,50 @@ import (
+		if len(lbls) > 0 {
+			params := endpointApi.NewGetEndpointsParams().WithLabels(lbls)
+			if e, err := client.Endpoint.GetEndpoints(params); err != nil {
+				Fatalf("Cannot get endpoint for given list of labels %s: %s\n", lbls, err)
```
"endpoint", as a single endpoint? We can have multiple endpoints using the same list of labels.

ianvernon replied:
Done.

aanm commented on daemon/endpoint.go:
```diff
@@ -28,7 +29,10 @@ import (
+	"github.com/docker/engine-api/types"
+	"github.com/docker/engine-api/types/filters"
 	"github.com/go-openapi/runtime/middleware"
+	ctx "golang.org/x/net/context"
```
can't you replace with `ctx "context"`?

ianvernon replied:
Done.

aanm commented on daemon/endpoint.go:
```diff
@@ -99,6 +103,48 @@ func (h *getEndpointID) Handle(params GetEndpointIDParams) middleware.Responder
+	}
+
+	containers, err := h.d.dockerClient.ContainerList(ctx.Background(), types.ContainerListOptions{
+		Size:   true,
```
Do you need the Size of the containers?

ianvernon replied:
I have removed interacting with the Docker client in my most recent upload.

aanm commented on daemon/endpoint.go:
```diff
@@ -99,6 +103,48 @@ func (h *getEndpointID) Handle(params GetEndpointIDParams) middleware.Responder
+
+	containers, err := h.d.dockerClient.ContainerList(ctx.Background(), types.ContainerListOptions{
+		Size:   true,
+		All:    true,
```
`All` includes dead containers, not sure if we need it as well.

ianvernon replied:
I have removed interacting with the Docker client in my most recent upload.

aanm commented on cilium/cmd/config.go:
```diff
@@ -75,6 +75,8 @@ func configDaemon(cmd *cobra.Command, opts []string) {
 
```
Why part of the `cilium config` command? I think it should make more sense as part of the `cilium status` on the `K8s api server`

ianvernon replied:
I had added this change initially as part of this commit, but ended up not needing it. Figured it might be useful if we added more visibility into what the configuration of Cilium is as part of the API, but I'm open to removing it as part of this patch. What do you think?

ianvernon replied:
Removed this in most recent upload.

Commit:
api: add pod / container name, labels for GET ep

* add new fields to Endpoint (pod-name, container-name)
* GET /endpoint now takes an optional array of labels to return a list of endpoints whose labels match the provided array.

Signed-off by: Ian Vernon <ian@covalent.io>

Commit:
cmd: get ep with labels, pod / container name, ID

* daemon: add handlers for new /endpoints API, update endpoint structures with container ID / pod name
* pkg/endpoint: support new API models
* pkg/endpointmanager: add pod name, container name entries as keys mapping to endpoints
* pkg/labels: change Orchestration labels -> OrchestrationIdentity labels (labels used in computing a security identity) and OrchrestationInfo labels (labels not used in computing a security identity)

`cilium endpoint get` now supports the following:
* `cilium endpoint get -l <set of labels>`
* `cilium endpoint get <eID, pod name, container name, etc.>`

Signed-off by: Ian Vernon <ian@covalent.io>
